src/crypto/common.rs

```diff
@@ -0,0 +1,256 @@
+
+    /// Verify a signature.
+    pub fn verify(
+        &self,
+        signed_data: &[u8],
+        signature: &[u8],
+    ) -> Result<(), AlgorithmError> {
+        match self {
+            #[cfg(feature = "ring")]
+            PublicKey::Ring(public_key) => {
+                public_key.verify(signed_data, signature)
+            }
+            #[cfg(feature = "openssl")]
+            PublicKey::Openssl(public_key) => {
+                public_key.verify(signed_data, signature)
+            }
+        }
+    }
+}
+
+/// Return the RSA exponent and modulus components from DNSKEY record data.
+pub fn rsa_exponent_modulus(
+    dnskey: &Dnskey<impl AsRef<[u8]>>,
+    min_len: usize,
+) -> Result<(Vec<u8>, Vec<u8>), AlgorithmError> {
+    let public_key = dnskey.public_key().as_ref();
+    if public_key.len() <= 3 {
+        return Err(AlgorithmError::InvalidData);
+    }
+
+    let (pos, exp_len) = match public_key[0] {
+        0 => (
+            3,
+            (usize::from(public_key[1]) << 8) | usize::from(public_key[2]),
+        ),
+        len => (1, usize::from(len)),
+    };
+
+    // Check if there's enough space for exponent and modulus.
+    if public_key[pos..].len() < pos + exp_len {
+        return Err(AlgorithmError::InvalidData);
+    };
+
+    // Check for minimum supported key size
+    if public_key[pos..].len() < min_len {
+        return Err(AlgorithmError::Unsupported);
+    }
+
+    let (e, n) = public_key[pos..].split_at(exp_len);
+    Ok((e.to_vec(), n.to_vec()))
+}
+
+//------------ AlgorithmError ------------------------------------------------
+
+/// An algorithm error during verification.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum AlgorithmError {
+    /// Unsupported algorithm.
+    Unsupported,
+
+    /// Bad signature.
+    BadSig,
+
+    /// Invalid data.
+    InvalidData,
+}
+
+//--- Display, Error
+
+impl fmt::Display for AlgorithmError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(match self {
+            AlgorithmError::Unsupported => "unsupported algorithm",
+            AlgorithmError::BadSig => "bad signature",
+            AlgorithmError::InvalidData => "invalid data",
+        })
+    }
+}
+
+impl error::Error for AlgorithmError {}
+
+//----------- FromDnskeyError ------------------------------------------------
+
+/// An error in reading a DNSKEY record.
+#[derive(Clone, Debug)]
+pub enum FromDnskeyError {
+    /// The key's algorithm is not supported.
+    UnsupportedAlgorithm,
+
+    /// The key's protocol is not supported.
+    UnsupportedProtocol,
+
+    /// The key is not valid.
+    InvalidKey,
+}
+
+//--- Display, Error
+
+impl fmt::Display for FromDnskeyError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::UnsupportedAlgorithm => "unsupported algorithm",
+            Self::UnsupportedProtocol => "unsupported protocol",
+            Self::InvalidKey => "malformed key",
```

Comment:
I see that this has only been moved so is not new, but looking at it I wonder are "invalid" and "malformed" really the same thing? Something can be well formed but still invalid, right? So maybe it would be better to make these consistent, in which case what is it: invalid or malformed?